core/src/credential.rs

```diff
@@ -80,4 +84,134 @@ impl Credential {
+            "familyName": "SMITH",
+            "gender": "Male",
+            "image": "data:image/png;base64,iVBORw0KGgo...kJggg==",
+            "residentSince": "2015-01-01",
+            "lprCategory": "C09",
+            "lprNumber": "999-999-999",
+            "commuterClassification": "C1",
+            "birthCountry": "Bahamas",
+            "birthDate": "1958-07-17"
+            },
+        });
+
+        let ds = Credential::deserialize(expect.to_string());
+        if ds.is_ok() {
+            let vc = ds.unwrap().serialize();
+            assert_json_eq!(expect, vc);
+        } else {
+            assert!(false);
+        }
+        Ok(())
+    }
+}
+
+mod formatter_context {
+    use serde::{self, Deserialize, Deserializer, Serializer};
+    use std::string::String;
+
+    pub fn serialize<S>(ctx: &Vec<String>, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let ctx_vec = crate::CONTEXT_CREDENTIALS
+            .into_iter()
+            .map(|s| s.to_string());
+        serializer.collect_seq(ctx_vec)
+    }
+
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<Vec<String>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = Vec::<String>::deserialize(deserializer)?;
+        let s = crate::CONTEXT_CREDENTIALS
+            .into_iter()
+            .map(|s| s.to_string())
+            .collect();
+        Ok(s)
+    }
+}
+
+mod formatter_credential_type {
+    use serde::{self, Deserialize, Deserializer, Serializer};
+
+    pub fn serialize<S>(cr_type: &String, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.collect_seq(cr_type.split(","))
+    }
+
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<String, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = Vec::<String>::deserialize(deserializer)?.join(",");
+        Ok(s)
+    }
+}
+
+mod formatter_credential_date {
+    use chrono::{DateTime, Utc};
+    use serde::{self, Deserialize, Deserializer, Serializer};
+    use std::string::String;
+    use std::time::SystemTime;
+
+    const FORMAT: &'static str = "%Y-%m-%dT%H:%M:%SZ";
+
+    pub fn serialize<S>(date: &SystemTime, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let formatted = DateTime::<Utc>::from(*date).format(FORMAT);
+        let s = formatted.to_string();
+        let s = format!("{}", &s);
+        serializer.serialize_str(&s)
+    }
+
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<SystemTime, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        let idate = DateTime::parse_from_rfc3339(&s);
+        if idate.is_ok() {
+            let sys_time = SystemTime::from(idate.unwrap());
+            return Ok(sys_time.clone());
+        } else {
+            return Err(serde::de::Error::custom(idate.unwrap_err()));
+        }
+    }
```

Comment:
For the error that can also be greatly simplified:
```suggestion
pub fn deserialize<'de, D>(deserializer: D) -> Result<SystemTime, D::Error>
    where
        D: Deserializer<'de>,
    {
        use serde::de::Error;
        let s = String::deserialize(deserializer)?;
        let idate = DateTime::parse_from_rfc3339(&s)
            .map(SystemTime::from)
            .map_err(Error::custom)?;
        Ok(idate)
    }
```

I’m using `.map` and `.map_err` to pass in functions directly.
I don’t need a closure since `.map` expects a function of some kind  `SystemTime::from` as well as `|d| SystemTime::from(d)` are identical in that respect